File: backend/services/item_service.py
"""
Item template service.

Handles item template persistence for worlds.
Items are stored in the items/ directory as the source of truth.
Player inventory (player.yaml) only stores references to items.
"""

import logging
import os
from dataclasses import dataclass
from typing import Any, Dict, List, Optional

import yaml

from services.world_service import WorldService

logger = logging.getLogger("ItemService")


@dataclass
class CachedItemTemplates:
    """Cache entry with directory mtime for invalidation."""

    templates: Dict[str, Dict[str, Any]]
    mtime: float


# Module-level cache
_item_templates_cache: Dict[str, CachedItemTemplates] = {}


class ItemService:
    """Item template management service.

    Items/ directory is the source of truth for all item definitions.
    player.yaml inventory only stores references (item_id, quantity, instance_properties).
    """

    @classmethod
    def _get_items_dir_mtime(cls, world_name: str) -> float:
        """Get the latest mtime of the items directory and its contents."""
        world_path = WorldService.get_world_path(world_name)
        items_dir = world_path / "items"

        if not items_dir.exists():
            return 0.0

        # Get max mtime of directory and all yaml files
        max_mtime = os.path.getmtime(items_dir)
        for item_file in items_dir.glob("*.yaml"):
            file_mtime = os.path.getmtime(item_file)
            max_mtime = max(max_mtime, file_mtime)

        return max_mtime

    @classmethod
    def save_item_template(
        cls,
        world_name: str,
        item_id: str,
        name: str,
        description: Optional[str] = None,
        properties: Optional[Dict[str, Any]] = None,
        overwrite: bool = False,
    ) -> bool:
        """
        Save an item template to items/{item_id}.yaml.

        Args:
            world_name: Name of the world
            item_id: Unique item identifier (used as filename)
            name: Display name of the item
            description: Item description
            properties: Item properties (stored as default_properties)
            overwrite: If True, overwrite existing template

        Returns:
            True if saved, False if already exists and overwrite=False
        """
        world_path = WorldService.get_world_path(world_name)
        items_dir = world_path / "items"

        # Ensure items directory exists
        items_dir.mkdir(exist_ok=True)

        # Sanitize item_id for filename
        safe_id = "".join(c for c in item_id if c.isalnum() or c in "._-")
        item_file = items_dir / f"{safe_id}.yaml"

        # Don't overwrite existing templates unless explicitly requested
        if item_file.exists() and not overwrite:
            logger.debug(f"Item template '{item_id}' already exists, skipping")
            return False

        template = {
            "id": item_id,
            "name": name,
            "description": description or "",
            "default_properties": properties or {},
        }

        with open(item_file, "w", encoding="utf-8") as f:
            yaml.dump(template, f, allow_unicode=True, default_flow_style=False)

        # Invalidate cache
        if world_name in _item_templates_cache:
            del _item_templates_cache[world_name]

        logger.info(f"Saved item template '{item_id}' in world '{world_name}'")
        return True

    @classmethod
    def load_item_template(cls, world_name: str, item_id: str) -> Optional[Dict[str, Any]]:
        """
        Load an item template from items/{item_id}.yaml.

        Uses cached templates for efficiency.

        Args:
            world_name: Name of the world
            item_id: Item identifier

        Returns:
            Item template dict if found, None otherwise
        """
        templates = cls.load_all_item_templates(world_name)
        return templates.get(item_id)

    @classmethod
    def load_all_item_templates(cls, world_name: str) -> Dict[str, Dict[str, Any]]:
        """
        Load all item templates for a world with mtime-based caching.

        Returns:
            Dict of item_id -> template data
        """
        current_mtime = cls._get_items_dir_mtime(world_name)

        # Check cache
        if world_name in _item_templates_cache:
            cached = _item_templates_cache[world_name]
            if cached.mtime >= current_mtime:
                logger.debug(f"ItemTemplates CACHE HIT for '{world_name}'")
                return cached.templates

        # Cache miss or stale - read from disk
        logger.debug(f"ItemTemplates CACHE MISS for '{world_name}' - reading from disk")

        world_path = WorldService.get_world_path(world_name)
        items_dir = world_path / "items"

        if not items_dir.exists():
            _item_templates_cache[world_name] = CachedItemTemplates(templates={}, mtime=current_mtime)
            return {}

        templates = {}
        for item_file in items_dir.glob("*.yaml"):
            try:
                with open(item_file, "r", encoding="utf-8") as f:
                    data = yaml.safe_load(f)
                    if data and "id" in data:
                        templates[data["id"]] = data
            except Exception as e:
                logger.warning(f"Failed to load item template {item_file}: {e}")

        # Update cache
        _item_templates_cache[world_name] = CachedItemTemplates(templates=templates, mtime=current_mtime)
        return templates

    @classmethod
    def resolve_inventory(
        cls,
        world_name: str,
        inventory_refs: List[Dict[str, Any]],
        normalize_props: bool = True,
    ) -> List[Dict[str, Any]]:
        """
        Resolve inventory references to full item data.

        Merges item templates from items/ with instance data from player.yaml.
        Supports both reference format (item_id + instance_properties) and
        legacy embedded format (full item data).

        Args:
            world_name: Name of the world
            inventory_refs: List of inventory entries from player.yaml
            normalize_props: If True, normalize properties to include higher_is_better metadata

        Returns:
            List of resolved inventory items with full data
        """
        from domain.services.player_rules import InventoryItem, normalize_properties

        templates = cls.load_all_item_templates(world_name)
        resolved = []

        for ref in inventory_refs:
            item_id = ref.get("item_id") or ref.get("id", "")
            template = templates.get(item_id)

            # Use from_reference to handle both formats
            item = InventoryItem.from_reference(ref, template)
            item_dict = item.to_dict()

            # Normalize properties to include higher_is_better metadata
            if normalize_props and item_dict.get("properties"):
                item_dict["properties"] = normalize_properties(item_dict["properties"])

            resolved.append(item_dict)

        return resolved

    @classmethod
    def to_reference_format(
        cls,
        world_name: str,
        inventory_items: List[Dict[str, Any]],
    ) -> List[Dict[str, Any]]:
        """
        Convert inventory items to reference format for saving.

        Also ensures templates exist in items/ directory.

        Args:
            world_name: Name of the world
            inventory_items: List of full inventory item dicts

        Returns:
            List of reference-format dicts (item_id, quantity, instance_properties)
        """
        from domain.services.player_rules import InventoryItem

        references = []
        templates = cls.load_all_item_templates(world_name)

        for item_data in inventory_items:
            item = InventoryItem.from_dict(item_data)

            # Ensure template exists
            if item.id not in templates:
                cls.save_item_template(
                    world_name,
                    item_id=item.id,
                    name=item.name,
                    description=item.description,
                    properties=item.properties,
                )

            references.append(item.to_reference_dict())

        return references

    @classmethod
    def get_all_items_in_world(cls, world_name: str) -> List[Dict[str, Any]]:
        """
        Get all item definitions in a world (from items/ directory).

        Returns:
            List of item template dicts
        """
        templates = cls.load_all_item_templates(world_name)
        return list(templates.values())

    @classmethod
    def delete_item_template(cls, world_name: str, item_id: str) -> bool:
        """
        Delete an item template from items/ directory.

        Args:
            world_name: Name of the world
            item_id: Item identifier to delete

        Returns:
            True if deleted, False if not found
        """
        world_path = WorldService.get_world_path(world_name)
        safe_id = "".join(c for c in item_id if c.isalnum() or c in "._-")
        item_file = world_path / "items" / f"{safe_id}.yaml"

        if not item_file.exists():
            return False

        item_file.unlink()

        # Invalidate cache
        if world_name in _item_templates_cache:
            del _item_templates_cache[world_name]

        logger.info(f"Deleted item template '{item_id}' from world '{world_name}'")
        return True

File: backend/src/services/item-service.ts
/**
 * Item templates — `worlds/{name}/items/{item_id}.yaml`, the source of truth for
 * what an item *is*; `player.yaml`'s inventory holds only references. Two
 * load-bearing behaviours: {@link ItemService.toReferenceFormat} writes
 * templates as a side effect, which is how an item invented mid-turn becomes
 * persistent; and templates are keyed by the file's own `id` field, not by its
 * filename, which is only a sanitised approximation of it.
 */

import { existsSync, mkdirSync, readFileSync, readdirSync, statSync, writeFileSync } from 'node:fs'
import { join } from 'node:path'
import { parse as parseYaml, stringify as stringifyYaml } from 'yaml'

import { MtimeCache, WorldService } from './world-service'
import { InventoryItem, normalizeProperties } from '../domain/player-rules'
import type { Equippable, InventoryEntry, ItemTemplate } from '../domain/player-rules'
import { getLogger } from '../infrastructure/logging/logger'

const logger = getLogger('ItemService')

export interface SaveItemTemplateInput {
  itemId: string
  name: string
  description?: string | null
  /** Written as `default_properties`; instances override individual keys. */
  properties?: Record<string, unknown> | null
  category?: string | null
  tags?: string[] | null
  rarity?: string | null
  icon?: string | null
  stacking?: Record<string, unknown> | null
  equippable?: Equippable | null
  usable?: Record<string, unknown> | null
  /** Without this an existing file is left alone and `false` is returned. */
  overwrite?: boolean
}

interface CachedTemplates {
  templates: Record<string, ItemTemplate>
  mtimeMs: number
}

// Narrower than the world-name sanitiser in `world-service.ts`, which also
// keeps the space — the two must not be shared.
const UNSAFE_ITEM_ID_CHARS = /[^\p{L}\p{N}._-]/gu

function isRecord(value: unknown): value is Record<string, unknown> {
  return typeof value === 'object' && value !== null && !Array.isArray(value)
}

function isNonEmpty(value: Record<string, unknown> | unknown[] | null | undefined): boolean {
  if (value === null || value === undefined) return false
  return Array.isArray(value) ? value.length > 0 : Object.keys(value).length > 0
}

// Item ids come from model output, so a bare `templates[id]` would resolve
// `"constructor"` against `Object.prototype`.
function getTemplate(
  templates: Record<string, ItemTemplate>,
  itemId: string,
): ItemTemplate | undefined {
  return Object.hasOwn(templates, itemId) ? templates[itemId] : undefined
}

function dumpYaml(data: unknown): string {
  return stringifyYaml(data, { sortMapEntries: true, indentSeq: false })
}

export class ItemService {
  private readonly worlds: WorldService

  // Not an MtimeCache: invalidation is the *maximum* mtime across `items/` and
  // every file in it, so add/remove/edit are all noticed.
  private readonly templateCache = new Map<string, CachedTemplates>()

  constructor(worldsDir: string, cache: MtimeCache = new MtimeCache()) {
    this.worlds = new WorldService(worldsDir, cache)
  }

  private itemsDir(worldName: string): string {
    return join(this.worlds.getWorldPath(worldName), 'items')
  }

  clearCache(): void {
    this.templateCache.clear()
  }

  private listTemplateFiles(itemsDir: string): string[] {
    let names: string[]
    try {
      names = readdirSync(itemsDir)
    } catch {
      return []
    }
    return names
      .filter((name) => name.endsWith('.yaml') && !name.startsWith('.'))
      .map((name) => join(itemsDir, name))
  }

  // The directory mtime catches an added or deleted file; each file's catches
  // an edit in place.
  private getItemsDirMtime(worldName: string): number {
    const itemsDir = this.itemsDir(worldName)

    let maxMtime: number
    try {
      maxMtime = statSync(itemsDir).mtimeMs
    } catch {
      return 0
    }

    for (const file of this.listTemplateFiles(itemsDir)) {
      try {
        maxMtime = Math.max(maxMtime, statSync(file).mtimeMs)
      } catch {
        // Deleted between listing and stat; the directory mtime already moved.
      }
    }

    return maxMtime
  }

  /**
   * Every template in a world, keyed by the `id` inside each file. An
   * unparseable or id-less file is skipped: one hand-edited template must not
   * take the whole inventory down mid-turn.
   */
  loadAllItemTemplates(worldName: string): Record<string, ItemTemplate> {
    const currentMtime = this.getItemsDirMtime(worldName)

    // Exact equality, not `>=`: a directory restored from an older copy must
    // reload rather than read as fresh.
    const cached = this.templateCache.get(worldName)
    if (cached && cached.mtimeMs === currentMtime) return cached.templates

    const templates: Record<string, ItemTemplate> = {}
    for (const file of this.listTemplateFiles(this.itemsDir(worldName))) {
      try {
        const data: unknown = parseYaml(readFileSync(file, 'utf-8'))
        if (isRecord(data) && 'id' in data) {
          templates[String(data.id)] = data as ItemTemplate
        }
      } catch (error) {
        logger.warning(`Failed to load item template ${file}: ${String(error)}`)
      }
    }

    this.templateCache.set(worldName, { templates, mtimeMs: currentMtime })
    return templates
  }

  loadItemTemplate(worldName: string, itemId: string): ItemTemplate | null {
    return getTemplate(this.loadAllItemTemplates(worldName), itemId) ?? null
  }

  /** Every item definition in a world, for the `/state/items` endpoint. */
  getAllItemsInWorld(worldName: string): ItemTemplate[] {
    return Object.values(this.loadAllItemTemplates(worldName))
  }

  /**
   * Write `items/{safe_id}.yaml`. Returns `false` — not an error — when the file
   * exists and `overwrite` is not set, so `persist_item` skips an item the world
   * already defines. `default_properties` is always written, even empty, because
   * the resolver distinguishes "no defaults" from a file predating it.
   */
  saveItemTemplate(worldName: string, input: SaveItemTemplateInput): boolean {
    const itemsDir = this.itemsDir(worldName)
    mkdirSync(itemsDir, { recursive: true })

    const safeId = input.itemId.replace(UNSAFE_ITEM_ID_CHARS, '')
    const itemFile = join(itemsDir, `${safeId}.yaml`)

    if (existsSync(itemFile) && !input.overwrite) {
      logger.debug(`Item template '${input.itemId}' already exists, skipping`)
      return false
    }

    // Build order is cosmetic — `dumpYaml` sorts keys on the way out.
    const template: ItemTemplate = {
      id: input.itemId,
      name: input.name,
      description: input.description || '',
    }

    if (input.category) template.category = input.category
    if (isNonEmpty(input.tags)) template.tags = input.tags
    if (input.rarity) template.rarity = input.rarity
    if (input.icon) template.icon = input.icon
    if (isNonEmpty(input.stacking)) template.stacking = input.stacking
    if (isNonEmpty(input.equippable)) template.equippable = input.equippable
    if (isNonEmpty(input.usable)) template.usable = input.usable

    template.default_properties = input.properties ?? {}

    writeFileSync(itemFile, dumpYaml(template), 'utf-8')

    // Makes a same-millisecond write visible where mtime resolution is coarser
    // than the gap between two saves.
    this.templateCache.delete(worldName)

    logger.info(`Saved item template '${input.itemId}' in world '${worldName}'`)
    return true
  }

  /**
   * Merge `player.yaml`'s inventory references with the templates they name,
   * instance properties over template defaults. An entry with no template still
   * resolves, through the legacy embedded path in `InventoryItem.fromReference`.
   */
  resolveInventory(
    worldName: string,
    inventoryRefs: InventoryEntry[],
    normalizeProps = true,
  ): InventoryEntry[] {
    const templates = this.loadAllItemTemplates(worldName)
    const resolved: InventoryEntry[] = []

    for (const ref of inventoryRefs) {
      const itemId = ref.item_id || ref.id || ''
      const itemDict = InventoryItem.fromReference(ref, getTemplate(templates, itemId)).toDict()

      if (normalizeProps && isNonEmpty(itemDict.properties)) {
        itemDict.properties = normalizeProperties(itemDict.properties)
      }

      resolved.push(itemDict)
    }

    return resolved
  }

  /**
   * Reduce full inventory items to the reference form `player.yaml` stores,
   * creating any missing template. The write is the point: an item conjured
   * mid-turn gets a durable definition at the moment the save file points at it.
   */
  toReferenceFormat(worldName: string, inventoryItems: InventoryEntry[]): InventoryEntry[] {
    const templates = this.loadAllItemTemplates(worldName)
    const references: InventoryEntry[] = []

    for (const itemData of inventoryItems) {
      const item = InventoryItem.fromDict(itemData)

      if (!Object.hasOwn(templates, item.id)) {
        this.saveItemTemplate(worldName, {
          itemId: item.id,
          name: item.name,
          description: item.description,
          properties: item.properties,
        })
      }

      references.push(item.toReferenceDict())
    }

    return references
  }
}
